haskell-best-practices.md

```diff
@@ -473,6 +473,214 @@ increases boilerplate.
 Add `LANGUAGE StrictData` to modules that contain only plain records that will
 undergo de/serialization (JSON, CSV, Database, etc).
 
+## Testing
+
+### Export Lists
+
+```hs
+-- Bad
+module FooSpec where
+
+-- Good
+module FooSpec
+  ( spec
+  ) where
+```
+
+### Imports
+
+Each project should have a `TestImport` used for all specs. Optionally, a
+`TestImport.IO` version can exist for writing non-App specs in `IO`.
+
+```hs
+-- Bad
+module Project.FooSpec
+  ( spec
+  ) where
+
+import Project.Prelude
+
+import Project.App
+import Project.Lib
+import TestImport
+
+-- Also Bad
+module Project.FooSpec
+  ( spec
+  ) where
+
+import Test.Hspec
+
+import Project.Prelude
+import Project.App
+import Project.Lib
+
+-- Good
+module Project.FooSpec
+  ( spec
+  ) where
+
+import TestImport
+
+import Project.App
+
+spec :: Spec
+spec = withApp loadApp $ do
+  describe "myAppFunction" $ do
+
+-- Also Good
+module Project.FooSpec
+  ( spec
+  ) where
+
+import TestImport.IO
+
+import Project.Lib
+
+spec :: Spec
+spec = do
+  describe "myPureFunction" $ do
+```
+
+#### CallStack
+
+Hspec uses the `HasCallStack` to report the location of test failures. It's
+important that any functions that make assertions (or call functions that make
+assertions) have the proper constraint.
+
+It's also important that the top-level `spec` **does not**. Otherwise, failures
+are attributed to the generated `Spec.hs` file.
+
+```hs
+-- Bad
+shouldApproximate :: Double -> Double -> Expectation
+
+-- Good
+shouldApproximate :: HasCallStack => Double -> Double -> Expectation
+
+-- Best (see below)
+shouldApproximate
+  :: (HasCallStack, MonadIO m)
+  => Double
+  -> Double
+  -> m ()
+```
+
+```hs
+-- Bad
+spec :: HasCallStack => Spec
+spec = do
+
+-- Good
+spec :: Spec
+spec = do
+```
+
+### Expectations vs Assertions
+
+Prefer Expectations over Assertions
+
+```hs
+-- Bad
+assertEqual "..." a b
+
+-- Good
+a `shouldBe` b
+```
+
+Except for functions that return values as part of their assertions, like
+`assertJust`, or `assertRight`.
```

Comment:
Should we settle on the use of these things vs e.g. `Just thingy <- ...`?

Reply:
We also have like `shouldBeJust` floating around. What should we do with those?

Reply:
`assertJust` and `Just x <-` are the same in that you get the `x`, but `shouldBeJust` is different as it's an expectation and doesn't return the `x`: ``shouldBeJust = (`shouldSatisfy` isJust) :: m ()``.

(If I'm wrong and `shouldBeJust` actually exists in a form that _does_ return the `x`, well then it should be renamed/replaced by `assertJust`.)

As for `assertJust` vs a pattern-match fail, I'm always down to pick a One True Way, but the usage feels about 50/50 in our code-bases. Dare you take a stand?